File: permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    
    const backtrack = (nums,currentPerm) => {
        if (nums.length == 0){
            result.push([...currentPerm]);
            return;
        }
        for (let x=0; x<nums.length; x++){
            const remaining = nums.slice(0,x).concat(nums.slice(x+1));
            backtrack(remaining, currentPerm.concat([nums[x]]));
        }
    }
    
    let result = [];
    backtrack(nums,[]);
    return result;
};

File: permutations.py
class Solution:
    def __init__(self):
        self.result = []

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.backtrack(nums,[]) 
        return self.result
    
    def backtrack(self,nums,current_perm):
        if not nums: # when no element in nums remaining 
            self.result.append(current_perm) # append the current permutation to result
            return
        for x in range(len(nums)): # for every element in nums
            self.backtrack(nums[:x]+nums[x+1:], current_perm+[nums[x]])
            # (rest of elements in nums except x, add current element x in current permuation)
